core/src/main/java/org/elasticsearch/search/aggregations/support/MultiValuesSourceAggregatorBuilder.java

```diff
@@ -0,0 +1,465 @@
+        }
+        this.valueType = valueType;
+        return (AB) this;
+    }
+
+    /**
+     * Gets the {@link ValueType} for the value produced by this aggregation
+     */
+    public ValueType valueType() {
+        return valueType;
+    }
+
+    /**
+     * Sets the format to use for the output of the aggregation.
+     */
+    @SuppressWarnings("unchecked")
+    public AB format(String format) {
+        if (format == null) {
+            throw new IllegalArgumentException("[format] must not be null: [" + name + "]");
+        }
+        this.format = format;
+        return (AB) this;
+    }
+
+    /**
+     * Gets the format to use for the output of the aggregation.
+     */
+    public String format() {
+        return format;
+    }
+
+    /**
+     * Sets the value to use when the aggregation finds a missing value in a
+     * document
+     */
+    @SuppressWarnings("unchecked")
+    public AB missing(Object missing) {
+        if (missing == null) {
+            throw new IllegalArgumentException("[missing] must not be null: [" + name + "]");
+        }
+        this.missing = missing;
+        return (AB) this;
+    }
+
+    /**
+     * Gets the value to use when the aggregation finds a missing value in a
+     * document
+     */
+    public Object missing() {
+        return missing;
+    }
+
+    /**
+     * Sets the time zone to use for this aggregation
+     */
+    @SuppressWarnings("unchecked")
+    public AB timeZone(DateTimeZone timeZone) {
+        if (timeZone == null) {
+            throw new IllegalArgumentException("[timeZone] must not be null: [" + name + "]");
+        }
+        this.timeZone = timeZone;
+        return (AB) this;
+    }
+
+    /**
+     * Gets the time zone to use for this aggregation
+     */
+    public DateTimeZone timeZone() {
+        return timeZone;
+    }
+
+    @Override
+    protected final MultiValuesSourceAggregatorFactory<VS, ?> doBuild(AggregationContext context, AggregatorFactory<?> parent,
+                                                                      AggregatorFactories.Builder subFactoriesBuilder) throws IOException {
+        Map<String, ValuesSourceConfig<VS>> configs = resolveConfig(context);
+        MultiValuesSourceAggregatorFactory<VS, ?> factory = innerBuild(context, configs, parent, subFactoriesBuilder);
+        return factory;
+    }
+
+    protected Map<String, ValuesSourceConfig<VS>> resolveConfig(AggregationContext context) {
+        HashMap<String, ValuesSourceConfig<VS>> configs = new HashMap<>();
+        for (String field : fields) {
+            ValuesSourceConfig<VS> config = config(context, field, null);
+            configs.put(field, config);
+        }
+        for (Script script : scripts) {
+            ValuesSourceConfig<VS> config = config(context, null, script);
+            configs.put(script.getType().getParseField().getPreferredName(), config);
```

Comment:
This name won't work as I can specify multiple scripts of a single type (e.g. `inline`) and the `ParseField` name for them all will be `"inline"` so they will overwrite each other. I think we need to go back to the parser here and parse a `Map<String, Script>` so each script can have a name. In the request this would look like:

```
{
    ...
    "script": {
        "my_script_1": {
            "inline": "script contents",
            "lang": "expressions"
        },
        "my_script_2": "script contents",
        ...
    }
```
Note that scripts can either be a JSON object or a String. The `Script.parse()` method handles both cases.